iostat/iostat.go

```diff
@@ -70,29 +72,78 @@ func (iostat *IOSTAT) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricTy
 	if err != nil {
 		return nil, err
 	}
-	metrics := make([]plugin.MetricType, len(mts))
-	for i, m := range mts {
-
-		if v, ok := data[m.Namespace().String()]; ok {
-			metrics[i] = plugin.MetricType{
-				Namespace_: m.Namespace(),
-				Data_:      v,
-				Timestamp_: time.Now(),
+
+	metrics := []plugin.MetricType{}
+
+	for _, mt := range mts {
+		ns := mt.Namespace()
+		if len(ns) < 5 {
+			return nil, fmt.Errorf("Namespace length is too short (len = %d)", len(ns))
+		}
+
+		if ns[4].Value == "*" {
+			if ns[3].Value == deviceMetric {
+				for k, v := range data {
+					dev, err := extractFromNamespace(k, 4)
+					if err != nil {
+						return nil, err
+					}
+
+					ns[4].Value = dev
```

Comment:
Seems like you overwrite `*` with for instance `sda1` then in next loop overwrite `sda1` with `sda2` and so on. Because ns[4] is always referring to same memory element. And slice memory is not being copied when you create new metric types later.